### PR TITLE
Add deserialization

### PR DESCRIPTION
**Pull request by mara-schulke**
Still a bit wip, need to write more test and align serialization support and deserialization support for certain types

### CODE REVIEW

**Comment by mara-schulke** [comment]
So i guess its ready for review @d-e-s-o :) The current implementation supports deserialization of all types that didnt held any data during the initial serialization.

I also added support for serializing structs into their type names. Its up to you if its outside of the scope of this crate :slightly_smiling_face: 

Have a nice day and let me know what you think :cherry_blossom:

**Comment by d-e-s-o** [comment]
Thanks for the pull request! I will take a closer look later this week. I will say, though, that I'd prefer if you didn't reformat the entire crate and if you cleaned up the commit history properly to have commits that logically build on top of each other and that do not merely represent some-work-in progress state from some time down the line -- that makes it hard to follow what is going on.

**Comment by mara-schulke** [comment]
I understand you absolutely, my reason for splitting the lib.rs / reorganizing the entire crate was the length of the code needed to implement a proper deserialization and the need of a shared and more flexible error handling.

Regarding the serialization i tried to limit the changes to type support and the shared error handling.

Renaming the public crate functions can be reversed if you want, but is now compliant to serde conventions

**Comment by mara-schulke** [comment]
Well don't all commit histories represent some work in progress state which isn't necessarily releasable? 🌞 But if you have a hint on how I should squash it / change it I'll do it ofc :)

**Comment by mara-schulke** [comment]
Any news on this @d-e-s-o ?

**Comment by d-e-s-o** [comment]
> Any news on this @d-e-s-o ?

I apologize. I haven't found the time to look through it in detail. Do you need it urgently?

**Comment by mara-schulke** [comment]
It's okay, take your time :)
Just notify me if you have any cr 
Have a nice day!

**Comment by d-e-s-o** [comment]
Sorry again for the delay. I went through the pull request today but in this state I unfortunately have to pass. Your changes effectively constitute a rewrite of the crate (changing everything from style, layout, API to functionality). That in and off itself does not disqualify it, of course, but functionality aside (which seems reasonable, though I am somewhat torn on the [necessary] requirement of supporting only unit-variants for deserialization) it violates close to everything my code typically stands for:
- the history is quite frankly a mess:
  - fix up commits have no place in a pull request
  - commits should stick to having a single purpose
  - a proper commit description is mandatory
- you make random unrelated changes (adding `Cargo.lock`, changing the import style of existing code, questionable changes to the README, ...)

I started fixing it up, but it would just take me too long and I don't really see the point. At the same time, I don't really have the time or interest to provide micro-level guidance in these matters. One can argue that I should provide more guidance on what I expect in a pull request in general, and while I take that point, it's generally just not worth it for me to write up a contribution guide for each tiny repository I put up (though I could perhaps come up with general guidelines I can link everywhere).

In general, it is my understanding that most open source projects stick to a certain style (for a very broad definition of "style", including everything mentioned above [and probably more that I forgot]) and have the implicit or explicit assumption that contributions conform to it. Furthermore, it is also my believe that most projects value an a priori discussion before seeing large API changing features implemented (in which certainly some of the topics above could be touched on/inquired as well, if unclear).

I understand (and appreciate) that you invested time in coming up with these changes. Perhaps the easiest way forward for you is to publish your own crate containing the functionality you desire.

If you feel strongly about inclusion in this crate (I am honored!) then my recommendation would be to first address the issues alluded to above and then open a new pull request (if something I mentioned is unclear or does not make logical sense, then I am happy to help [and generally responsive]).

**Comment by mara-schulke** [comment]
Okay no problem.

I want to say that i think that this type of response to prs is neither decent nor welcoming. If you want your project to be _your_ project and in _your_ style, i can respect that. But in the end the reason i wanted to merge these changes into `serde_variant` is rooted in the fact that i want to prevent polluting the crates.io ecosystem with a bunch of crates that do roughly the same. There should be one advanced one instead of multiple smaller ones.

Furthermore i dont see your strict and opinionated requirements on the git history as a given for every pull request that is opened in the opensource world. That may be a good and tidy style, but in the end they are just that - style. In the end and i dont think a rather rude reply helps to enforce them.

Also i think pinning your project as an example for "the opensource rust style" is a bit pretentious, especially considering that you ignored quite a few idioms and conventions in your original implementation. This doesnt mean that i dont need to comply to that style, this is ultimately your project, but you should consider not seeing this as a given for other people in the rust ecosystem and be a bit more friendly to others.

Have a nice evening.
